File: h5gl/new/src/pan3d/display/particle/locus/Display3DLocusPartilce.ts
module Pan3d {
    export class Display3DLocusPartilce extends Display3DParticle {


        public get locusdata(): ParticleLocusData {
            return <ParticleLocusData>this.data;
        }
        public creatData(): void {
            this.data = new ParticleLocusData(this.scene3D);
        }
        public setVa(): void {

            var ctx:Context3D=this.scene3D.context3D;
            var tf: boolean =ctx.pushVa(this.data.objData.vertexBuffer);
            if (!tf) {
                ctx.setVaOffset(0, 3, this.data.objData.stride, 0);
                ctx.setVaOffset(1, 2, this.data.objData.stride, 28);
                if (this.data._watchEye) {
                    ctx.setVaOffset(2, 4, this.data.objData.stride, 12);
                }
            }
            this.setMaterialTexture();
            ctx.drawCall(this.data.objData.indexBuffer, this.data.objData.treNum);
        }

        public setVc(): void {
            this.updateUV();
            var ctx:Context3D=this.scene3D.context3D;
        
            this.data.vcmatData.set(this.scene3D.viewMatrx3D.m, 0);
    
            this.data.vcmatData.set(this.scene3D.cam3D.cameraMatrix.m, 16);
        
            this.data.vcmatData.set(this.modelMatrix.m, 32);
       
            this.data.vcmatData.set(this.locusdata._resultUvVec, 48);

            if (this.data._watchEye) {

                this.locusdata._caramPosVec[0] = this.scene3D.cam3D.x;
                this.locusdata._caramPosVec[1] = this.scene3D.cam3D.y;
                this.locusdata._caramPosVec[2] = this.scene3D.cam3D.z;
 
                this.data.vcmatData.set(this.locusdata._caramPosVec, 52);
            }

            if (this.locusdata._changUv) {
                this.data.setFloat32Vec("isUv", this.locusdata._uvVec);//56
            }

            ctx.setVcMatrix4fv(this.data.materialParam.shader, "vcmat", this.data.vcmatData);

            this.setMaterialVc();

        }

        public updateUV(): void {
            var $nowTime: number = this._time / Scene3D.frameTime;
            var $lifeRoundNum: number = (this.data._life / 100);
            var $moveUv: number = this.locusdata._speed * $nowTime / this.locusdata._density / 10
            if (this.locusdata._isEnd) {
                $moveUv = Math.min(1, $moveUv);
            }

            if (this.locusdata._isLoop) {
                if (this.locusdata._life) {
                    $moveUv = $moveUv % ($lifeRoundNum + 1)
                } else {
                    $moveUv = $moveUv % 1;
                }
            }

            this.locusdata._resultUvVec[0] = $moveUv;
        }


 

    }
}

File: metalios/iosgl/iosgl/pan3d/display/particle/locus/Display3DLocusPartilce.m
//
//  Display3DLocusPartilce.m
//  iosgl
//
//  Created by zhao on 25/3/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "Display3DLocusPartilce.h"
#import "ParticleLocusData.h"
#import "Display3DLocusShader.h"
#import "Context3D.h"
#import "Scene3D.h"
#import "Scene_data.h"
#import "ProgrmaManager.h"
#import "Display3DSprite.h"
#import "ParticleMetalType.h"
 
@implementation Display3DLocusPartilce


-(ParticleLocusData*)locusdata;
{
    return (ParticleLocusData*)self.data;
}
-(ObjData*)particleGpuObjData;
{
    return self.locusdata.objData;
}
 
-(void)onCreated;
{
    
}
- (void)update;
{
//      [super update];
    
    NSLog(@"规矩被屏蔽")
}
- (void)setVc;
{
    [self setViewCamModeMatr3d];
    
     
    Camera3D* cam3D=self.scene3D.camera3D;
    [self updateUV];
    Vector3D*  scaleVec =   self.locusdata._resultUvVec;
 
    Vector3D*  caramPosVec = [[Vector3D alloc]x:cam3D.x y:cam3D.y z:cam3D.z];
    if (self.data._watchEye) {
 
    }
 
  
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
    
 
    
    ParticleMetalLocusMatrixData matrixList = {[cam3D.viewMatrix getMatrixFloat4x4], [cam3D.camMatrix3D getMatrixFloat4x4], [self.modeMatrix getMatrixFloat4x4]};
  
   [renderEncoder setVertexBytes:&matrixList
                          length:sizeof(matrixList)
                         atIndex:3];
    
    ParticleMetalLocusVcmatData VcmatData = { (vector_float4){scaleVec.x,scaleVec.y,scaleVec.z,scaleVec.w} ,(vector_float4){caramPosVec.x,caramPosVec.y,caramPosVec.z,caramPosVec.w}};
     
   [renderEncoder setVertexBytes:&VcmatData
                          length:sizeof(VcmatData)
                         atIndex:4];
    
    
}
-(NSUInteger)getFcDataIdx;
{
    return 2; //FC数据的位置，不同特效不一样
}
-(void)updateUV;
{
  
    float nowTime=self._time/[Scene_data default].frameTime;
    float  lifeRoundNum=self.data._life / 100.0f;
    float moveUv = self.locusdata._speed * nowTime / self.locusdata._density / 10.0f;
    if (self.locusdata._isEnd) {
        moveUv = MIN(1, moveUv);
    }
    if (self.locusdata._isLoop) {
        if (self.locusdata._life) {
            moveUv = moveUv- ceilf(moveUv/(lifeRoundNum+1))*(lifeRoundNum+1) ;
        } else {
            moveUv = moveUv- ceilf(moveUv/1)*1 ;
        }
    }
    self.locusdata._resultUvVec.x = moveUv;
    
}
- (void)setVa;
{
 
    ObjData* temp=self.particleGpuObjData;
    
 
  
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
    
    
     [renderEncoder setVertexBuffer: temp.mtkvertices
                             offset:0
                            atIndex:0];
    [renderEncoder setVertexBuffer: temp.mtkuvs
                            offset:0
                           atIndex:1];
    [renderEncoder setVertexBuffer: temp.mtknrms
                            offset:0
                           atIndex:2];
    
    
    
   [renderEncoder drawIndexedPrimitives:MTLPrimitiveTypeTriangle
                             indexCount: temp.mtkindexCount
                              indexType:MTLIndexTypeUInt32
                            indexBuffer: temp.mtkindexs
                      indexBufferOffset:0];
    


}
- (void)resetVa;
{
      Context3D *ctx=self.scene3D.context3D;
    [ctx clearVa:0];
    [ctx clearVa:1];
    [ctx clearVa:2];
}
@end

